spec/ruby/core/random/default_spec.rb

```diff
@@ -1,8 +1,17 @@
 require_relative '../../spec_helper'
 
 describe "Random::DEFAULT" do
-  it "returns a Random instance" do
-    Random::DEFAULT.should be_an_instance_of(Random)
+
+  ruby_version_is ''...'3.0' do
+    it "returns a Random instance" do
+      Random::DEFAULT.should be_an_instance_of(Random)
+    end
+  end
+
+  ruby_version_is '3.0' do
+    it "returns a Random instance" do
+      Random::DEFAULT.should be_an_instance_of(Class)
+    end
```

Comment:
```suggestion
  it "returns a random number generator" do
    Random::DEFAULT.should respond_to(:rand)
  end
```

Reply:
It seems nice, but OTOH the current diff has the advantage that other Ruby implementations will notice the change and that they need to change what Random::DEFAULT points it when updating to Ruby 3.
Maybe have both?

Reply:
It can be anything which generates (pseudo) random numbers.

Reply:
ok, both.

Reply:
It doesn't feel reasonable that "the PRNG should be a class", but rather nonsense.
"It should respond to `rand`" would be fine as duck-typing.

Reply:
> It doesn't feel reasonable that "the PRNG should be a class", but rather nonsense.
> "It should respond to `rand`" would be fine as duck-typing.

I'll change it to `Random::DEFAULT.should.equal?(Random)` (for Ruby>=3), that's way the change done here is clear and not missed by other Ruby implementations. It might affect semantics in non-trivial ways.

Reply:
Do you expect the code like `Random::DEFAULT.new` in Ruby 3?
Otherwise these don't make sense.

Reply:
See https://github.com/ruby/ruby/commit/f0e2ea2859f9a51260f22b38fae1719a2eab59d9
Basically, an important part of ruby/spec is to track changes between versions, so alternative Ruby implementations can follow those changes for best compatibility.

Reply:
The compatibility users can expect is just that it works as a RNG.
Whether it is an instance of Random or itself is an implementation detail, and other implementations may select anything as it does.

Reply:
> other implementations may select anything as it does.

The reality is alternative Ruby implementations need to follow the CRuby behavior as close as possible, or they risk various compatibility issues. Some gems even depend on the sequence of Random for a given seed! (not a good idea, but still a fact). The class of Random::DEFAULT is something user-observable, so it should be tested.
I get it might not be part of a "formal specification of the Ruby language", but consider ruby/spec is basically a test suite that's shared between Ruby implementations and to which any Ruby implementation can contribute conveniently, and two of the main goals are to improve compatibility between Ruby implementations and to improve the coverage of tests for Ruby behavior.

Reply:
It is an implementation detail which is useless to be shared.
In fact, the constant might be a proxy object, and could be changed.
What we consider is just that "it works as a RNG".

Reply:
Till now, no code uses that "Random::DEFAULT equals Random", of course.
If we publish this statement, users may write such code and we (including other implementors) will have to accept it.
If we don't, we can reject such code.
This is chicken-and-egg.

Reply:
Because `Random::DEFAULT` is `Random` will be in 3.0, users might depend on it. It would be nicer if they don't , but they might and there is nothing we can do about it, except if we deprecate the Random::DEFAULT constant.
The fact there is a test in ruby/spec doesn't mean some code is supported, recommended or anything. It just tries to test the behavior, in as much details as possible to improve compatibility between Ruby implementations.
Probably lots of tests in ruby/spec are not what users should do, but they are useful tests nevertheless.

BTW, I think we should mention this change in `NEWS` because it could cause some incompatibility.

Reply:
> The fact there is a test in ruby/spec doesn't mean some code is supported, recommended or anything. It just tries to test the behavior, in as much details as possible to improve compatibility between Ruby implementations.
> Probably lots of tests in ruby/spec are not what user code should do, but they are useful tests nevertheless.

Probably I've misunderstood your intention, ruby/spec is only for other implementors, but the other users should not rely on it?

Reply:
Looks like I forgot to reply to this.
I think ruby/spec is primarily for Ruby implementors yes, similar to the CRuby test suite.
I don't think many Ruby users look at ruby/spec, although of course they enjoy better-tested Ruby implementations.